### DISCUSSION TITLE
Is there any way to create reusable functionality, like in Vue 3 composables?

### DISCUSSION THREAD

**Discussion by alterastro** [alpinejs/alpine#4224] [1. Help]
I'm developing a set of UI components using Alpine.js, and I absolutely miss the ability to create functionality that is reusable between components.

Is it possible to create reusable functionality similar to the Vue.js Composition API?
(The ability to create something like mixins from Vue 2 is not suitable and it’s terrible)

I know that many people create headless components, and when implementing them they use the full capabilities of reactivity, but this is not exactly what is needed, since it makes it inconvenient to interact with such components if they are more complex than a conventional accordion.

This may be related to the [discussion](https://github.com/alpinejs/alpine/discussions/4114#discussion-6432157).

Is there some way to implement this?

**Comment by ekwoka** [comment]
You'd really need to provide more info on exactly what you're trying to work with.

custom directives are a good way to encapsulate functionality. Is there a reason that doesn't work for you?

**Reply by alterastro** [reply]
Yes, just while I was writing this post, some ideas arose for implementing some functionality through directives.
There is one BUT: directives can accept and process reactive values, but cannot return new calculated reactive variables based on the input data (Correct me if I'm presumptuously wrong).

That is, for example, I need to reuse the functionality of the stack (for dialogs, tooltips, etc.). In Vue, I would create a composable that would take 2 reactive variables as input: isActive and zIndex. And at the output I would get 2 reactive variables: isTop and stackStyles. And in the component itself, for example, I can pass isTop to another composable, for example to create methods that respond to component events.
That is, I have full control and reusability of the functionality.

**Reply by SimoTod** [reply]
Directives add a behaviour to your html tag. You can define functions returning reactive values, they are called magics.

That being said, x-data and x-bind are usually enough to cover reusable components. 

Maybe try to post an example (like an html snippet showing what you want to achieve, it doesn't matter if it doesn't work)

**Reply by SimoTod** [reply]
Just re-read your previous post. What you want is for your directive to call Alpine.addScopeToNode passing the reactive output value to the $root. From that point, the following directives on the same component will have access to your new values. You'll need to make them reactive with Alpine.reactive before adding them to the node, I think (look at the source code for x-data)

**Reply by ekwoka** [reply]
They can do that.

They can add scope to the node itself that children can then access.

**Reply by alterastro** [reply]
I think we can start with a primitive example:
Let's say we have a certain component, for example a menu, which has a modelValue property, which we can set through the initial parameters in x -data.
Next I need some computed property like model. Naturally, for convenience, I implement it as a getter and setter in order to interact with it as with a regular property.
```javascript
Alpine.data('menu', (props) => ({
    modelValue: false,
    ...props,

    /* ... */
    get model() {
        return this.modelValue
    },
    set model(value) {
        /* There may be some complex logic here */
	this.$dispatch('update:model-value', value)
        this.modelValue = value
    },
    /* ... */
}))
```
The problem is that if I want to reuse the model logic, I won’t be able to do it, since I won’t be able to pass the context.
```javascript
/* It is also possible through an arrow function, but the essence does not change, we do not have access to this, which is logical. */
function useXModel(ctx) {
    return {
        get model() {
            return ctx.modelValue
        },
        set model(value) {
            /* There may be some complex logic here */
            this.$dispatch('update:model-value', value)
            ctx.modelValue = value
        },
    }
}

Alpine.data('menu', (props) => ({
    modelValue: false,
    ...props,

    /* ... */
    /* this will be undefined */
    ...useXModel(this),
    /* ... */
}))
```
In general, I would like to have something similar to the code above. Perhaps there is another workaround?

**Reply by ekwoka** [reply]
You can do that just like in the solution demonstrated here:

https://github.com/alpinejs/alpine/pull/4200#issuecomment-2099749022
